### ISSUE TITLE
[Feature Request]: MSBuild Server Identifier

### ISSUE THREAD

**Issue by hknielsen** [dotnet/msbuild#9692]
### Summary

When having multiple Clones of the same repository using MSBuild Server for speeding up compilation, we end up in situations where the MSBuild Server gets shutdown from the other clone, while a compilation is ongoing, resulting in errors because of Server Shutdown.

### Background and Motivation

When having multiple Clones of the same repository using MSBuild Server for speeding up compilation, we end up in situations where the MSBuild Server gets shutdown from the other clone, while a compilation is ongoing, resulting in errors because of Server Shutdown.

### Proposed Feature

Add an Identifier to the started server + nodes started by the server, so repositories can't interfere with each other.
It would be optimal if:
- The Identifier could be passed as an argument or a new env var
- The Identifier would be additive to the current `CommunicationsUtilities.GetHandshakeOptions` so it would work as today if no extra identifier was added.
- Add the public api to `MSBuildClient` to support MSBuild Server Identifier.

This would 

### Alternative Designs

The MSBuild Server it identified with the location, so if we have two copies of MSBuild, each would have a unique identity and MSBuild Server + Nodes.

**Comment by hknielsen** [comment]
> Seems related to [dotnet/sdk#9303](https://github.com/dotnet/sdk/issues/9303) and [dotnet/sdk#20183](https://github.com/dotnet/sdk/issues/20183). Could this use the same `SharedCompilationId` property as Roslyn?

It would make alot of sense that all MSBuild nodes and Roslyn would use the same Identifier, passing along the MSBuild SharedId to Roslyn.

**Comment by hknielsen** [comment]
@KalleOlaviNiemitalo updated the proposal to use the same `Shared-Id` between Roslyn and MSBuild

**Comment by JaynieBai** [comment]
@hknielsen sorry, I can't repro your issue as the following steps. Did you repro you issue in your specified repo? or is there anything that I missed?
1. Enable MSBuild Server with DOTNET_CLI_USE_MSBUILD_SERVER = 1
2. Clone two copies of msbuild repos and execute dotnet build MSBuild.sln
3. dotnet build-server shutdown when one build is done and the other is still going.

**Comment by hknielsen** [comment]
@JaynieBai We were seeing it randomly in CI. Basically it was related to rebuilds on multiple clones using the same msbuild server, where one finished and shutsdown the server.

**Comment by rainersigwald** [comment]
@hknielsen and to be clear: you are explicitly setting `DOTNET_CLI_USE_MSBUILD_SERVER` in your CI jobs?

**Comment by hknielsen** [comment]
@rainersigwald  - yeah, we are setting the env var to enable the MSBuild Servier. And calling shutdown when done. We were also not only using this on CI, so when we are done building, we cleanup resources, including shutdown the MSBuild Server

**Comment by baronfel** [comment]
Roslyn/Razor are also hitting similar issues with the Razor compiler server and the `build-server shutdown` command. cc @jaredpar - it would be cool if we could find a reusable solution for our different servers.

**Comment by jaredpar** [comment]
I agree that we need to find a reusable solution for our different servers. Right now we're all solving the same problem but in different ways which each have their own flaws.

**Comment by JanProvaznik** [comment]
Cross-referencing the broader MSBuild Server default-on analysis at #9379 (overnight investigation summary).

This issue is **Infrastructure / Class A** — cross-build server identity collision.

**Status update on related work:**
- **PR #13651** (which would have added `Path.GetTempPath()` to the non-TaskHost handshake salt and partially addressed your scenario for the multi-TMPDIR case) was **closed as "too heavy"** on 2026-05-05 — the temp-path fix did NOT land. So today the salt is still `hash(MSBUILDNODEHANDSHAKESALT + toolsDirectory)` only.
- The broader `SharedId` / workspace-isolation feature you're asking for here is therefore still entirely unaddressed.

**Today's workaround (no code change required):** set `MSBUILDNODEHANDSHAKESALT` to a unique per-workspace value (e.g., the repo name, a workspace path hash, or `git rev-parse --show-toplevel | sha256sum`) in your environment or per-invocation. This gives each workspace its own server pipe identity. The VMR is being recommended this exact approach (`MSBUILDNODEHANDSHAKESALT=$(RepositoryName)` per-repo `<Exec>`).

**The first-class `SharedId`/workspace API requested here would** promote that workaround to a documented field of the `ServerNodeHandshake` and let consumers (VS, the dotnet CLI, IDEs) compute it from project metadata rather than asking every user to set an env var. Worth the design work because the env-var workaround doesn't compose well with VS / multi-clone interactive scenarios where the user has no chance to set env per launch.

**Comment by KalleOlaviNiemitalo** [comment]
In the meantime, I've found that setting `SharedCompilationId` can cause conflicts if MSBuild is used on a solution in which different projects are compiled using different versions of Roslyn (e.g. one from Visual Studio and another from .NET SDK).  Fixing this may require changing `SharedCompilationId` per project,  but that shouldn't require using different MSBuild servers too, so I now think `SharedCompilationId` itself should not be used for selecting MSBuild servers.
